text/0000-try-from.md

```diff
@@ -0,0 +1,165 @@
+- Feature Name: try_from
+- Start Date: 2016-03-10
+- RFC PR:
+- Rust Issue:
+
+# Summary
+[summary]: #summary
+
+The standard library provides the `From` and `Into` traits as standard ways to
+convert between types. However, these traits only support *infallable*
+conversions. This RFC proposes the addition of `TryFrom` and `TryInto` traits
+to support these use cases in a standard way.
+
+# Motivation
+[motivation]: #motivation
+
+Fallible conversions are fairly common, and a collection of ad-hoc traits has
+arisen to support them, both [within the standard library][from-str] and [in
+third party crates][into-connect-params]. A standardized set of traits
+following the pattern set by `From` and `Into` will ease these APIs by
+providing a standardized interface as we expand the set of fallible
+conversions.
+
+One specific avenue of expansion that has been frequently requested is fallible
+integer conversion traits. Conversions between integer types may currently be
+performed with the `as` operator, which will silently truncate the value if it
+is out of bounds of the target type. Code which needs to down-cast values must
+manually check that the cast will succeed, which is both tedious and error
+prone. A fallible conversion trait reduces code like this:
+
+```rust
+let value: isize = ...;
+
+let value: u32 = if value < 0 || value > u32::max_value() as isize {
+    return Err(BogusCast);
+} else {
+    value as u32
+};
+```
+
+to simply:
+
+```rust
+let value: isize = ...;
+let value: u32 = try!(value.try_into());
```

Comment:
It's not all that uncommon in my experience:
https://github.com/sfackler/rust-postgres/blob/master/src/types/mod.rs#L700
https://github.com/sfackler/rust-postgres-large-object/blob/master/src/lib.rs#L220

But in any case, what you do with the `Err` once you get it doesn't really have much to do with this RFC - it's focused on how you get that `Err`.